chain/chain/src/resharding/trie_state_resharder.rs

```diff
@@ -0,0 +1,273 @@
+        } else {
+            store_update.set(
+                DBCol::Misc,
+                TRIE_STATE_RESHARDING_STATUS_KEY,
+                &borsh::to_vec(status)?,
+            );
+        }
+        store_update.commit()?;
+        Ok(())
+    }
+
+    fn load_status(&self) -> Result<Option<TrieStateReshardingStatus>, Error> {
+        Ok(self
+            .runtime
+            .store()
+            .get_ser::<TrieStateReshardingStatus>(DBCol::Misc, TRIE_STATE_RESHARDING_STATUS_KEY)?)
+    }
+
+    /// Start a resharding operation by iterating the memtries of each child shard,
+    /// writing the result to the `State` column of the respective shard.
+    pub fn start_resharding_blocking(
+        &self,
+        event: &ReshardingSplitShardParams,
+    ) -> Result<(), Error> {
+        if let Some(status) = self.load_status()? {
+            panic!(
+                "TrieStateReshardingStatus already exists for shard {}, cannot start a new resharding operation. Run resume_resharding to continue.",
+                status.parent_shard_uid
+            );
+        }
+
+        // Get state root from the chunk extra of the child shard.
+        let block_hash = event.resharding_block.hash;
+        let store = self.runtime.store().chain_store();
+        let left_state_root =
+            *store.get_chunk_extra(&block_hash, &event.left_child_shard)?.state_root();
+        let right_state_root =
+            *store.get_chunk_extra(&block_hash, &event.right_child_shard)?.state_root();
+        tracing::info!(
+            target: "resharding",
+            ?left_state_root,
+            ?right_state_root,
+            ?event.left_child_shard,
+            ?event.right_child_shard,
+            "TrieStateResharding: child state roots"
+        );
+
+        let mut status = TrieStateReshardingStatus::new(
+            event.parent_shard,
+            TrieStateReshardingChildStatus::new(event.left_child_shard, left_state_root),
+            TrieStateReshardingChildStatus::new(event.right_child_shard, right_state_root),
+        )
+        .with_metrics();
+        self.resharding_blocking_impl(&mut status)
+    }
+
+    /// Resume an interrupted resharding operation.
+    pub fn resume(&self, parent_shard_uid: ShardUId) -> Result<(), Error> {
+        let Some(status) = self.load_status()? else {
+            tracing::info!(target: "resharding", "Resharding status not found, nothing to resume.");
+            return Ok(());
+        };
+
+        if status.parent_shard_uid != parent_shard_uid {
+            return Err(Error::ReshardingError(format!(
+                "Resharding status shard UID {} does not match the provided shard UID {}.",
+                status.parent_shard_uid, parent_shard_uid
+            )));
+        }
+        let mut status = status.with_metrics();
+        self.resharding_blocking_impl(&mut status)
+    }
+
+    fn resharding_blocking_impl(
```

Comment:
Do we have this indirection and not a direct call to `process_batch_and_update_status` to add removing mapping in the future?

Reply:
mostly here for future unit tests.